### DISCUSSION TITLE
Pyramidの数値型の仕様案について

### DISCUSSION THREAD

**Discussion by Shinonome517** [tsumiki-works/Pyramid#121] [General]
# Pyramidの数値型の仕様案

## 提案

Pyramidの数値型は、

* int: 32bit符号付き整数
* float: 32bit符号付き浮動小数（IEEE754単精度浮動小数）

の二つにすることを提案する。

**Contributer の皆さんには**賛成意見**の表明、理由付きの**反対意見**の表明、**修正案**の提案、見落としている**問題点**の指摘などをお願いしたい**

## 理由（簡潔に）

* TypeScript（JavaScript）は数値型に`Number`型（64bit浮動小数）しか持たない
* 64bit整数（もしくはPythonのような任意精度整数）を実用に耐えうるレベルで用意することが困難でありかつデメリットも多い

### そもそも何を目的としているか？

Pyramidは、関数型パラダイムを支援するビジュアルプログラミング言語である。

* ビジュアルプログラミングは、不正な入力を弾きやすい
* 関数型パラダイムは数学表現と相性が良い（数値計算においても、バグが少ないコードを書きやすいという点でメリットがある）
* 教育の観点から、`int`と`float`の違いがあった方が良い

多少、個人の主観が入っているが、以上の理由から整数型と浮動小数型の二つは用意したいと考えている。

### TypeScriptの数値型の仕様について

以上の理想に反し、TypeScriptの数値型は`Number`型（中身は64bit浮動小数）しか存在しない。
このため、浮動小数演算で整数演算を再現する必要がある。

なお、後述の実験などからTypeScriptは、Pythonなどでは整数として扱われるリテラル（ex, 1234など）に対しても、浮動小数演算をおこなっている可能性が高いことがわかる。

### 64bit浮動小数で整数型を再現するときに起こる問題

私は、主に二点だと考えている。

* 丸め誤差の混入
* 情報落ち

現状、前者は大きな問題ではないと考えているが、後者は深刻な問題を孕む。

#### 丸め誤差の混入について

旧Pyramidの演算で生じていたように、浮動小数の演算では、演算の内容によっては浮動小数誤差が入る可能性がある。

```rust
fn main(){
    println!("{}", 1.32 + 2.08); //3.4000000000000004
}
```

つまり、浮動小数で整数の演算を行うのにはリスクが伴う。（この誤差が誤差が入る条件を調べていないため、小数点以下が全て0の演算ではこのような誤差が混入することはないのかもしれない）

但し、JavaScriptでの演算は`Number.MAX_SAFE_INTEGER`メソッドの存在からも分かるように、また、適宜`ParseInt()`などを用いることで十分小さな整数では正しく整数演算ができると考えている。

#### 情報落ち

絶対値の大きく異なる二数の加減算では、小さな方の値は無視されてしまう。

これが、浮動小数型で整数型を再現するときに大きな問題となる。

64bit浮動小数の仮数部は53bitであるので、 $\log_{10} 2^{53} \simeq 15$ から15桁ほどしか精度がないのだが、対して64bit型浮動小数は $\log_{10} 2^{64} \simeq 19$ 桁の精度が必要になる。

このため、64bit浮動小数では64bit整数を表現することができない。

実際16桁以上の値に1を加えると情報落ちする。

```TypeScript
let a = 10000000000_00000
let b = 10000000000_000000

console.log(a + 1) //1000000000000001
console.log(b + 1) //10000000000000000  (expected 10000000000000001)
```

#### BigInt型を用いる

JavaScriptには、任意精度の整数型を歌う[BigInt](https://ja.javascript.info/bigint)が存在するが、検証してみた結果それは嘘である。

<img width="451" alt="スクリーンショット 2022-11-23 22 22 02" src="https://user-images.githubusercontent.com/66119716/203592343-0d366a37-2de5-4a41-82b3-e04022eddb0a.png">
<img width="220" alt="スクリーンショット 2022-11-23 22 21 44" src="https://user-images.githubusercontent.com/66119716/203592408-48e0d1ad-2afe-48a5-8e04-6fb01d25dacd.png">

また、仮にこの型が任意精度であったとしても、`BigInt`型は`Number`型と併用することができない。
`BigInt`は`Nuber`と比べて演算が遅く、速度の観点から適切に`BigInt`と`Nuber`型を切り替える必要があり、煩わしい。
また、`BigInt`型は`Math`モジュールの組み込み関数（`sin`, `exp`, `floor`, `abs`, etc..）を使うことができないので、自前で実装する必要がある。
15桁から19桁までの整数のために、この労力は見合わないと判断した。

#### 任意精度整数演算をソフトウェア的に実装する

現状一番マシな選択肢である。
但し、@Shinonome517のキャパを全て食い尽くす。
また、`BigInt`と同様、組み込みの数学関数は使うことができないので、実装が大変である。

#### 演算のみJavaScriptからFFIでC++コードを呼ぶ

Node.jsを用いることで[FFIからC++を呼ぶことができるらしい](https://zenn.dev/kn64872/articles/dbb92605aec7bb)が、TypeScriptでできるか不明（要詳細調査）。
また、すべてをTypeScriptで統一した仕様変更に逆行しており、気が進まない。

#### JavaScriptエンジンのコードを読み整数演算ユニットを使用する方法を探す

[GoogleのJavaScriptエンジンv8](https://chromium.googlesource.com/v8/v8/)のソースコードを読もうとしたが、非常に難解で断念。アセンブリに変換しているコードから、（存在するかもわからない）整数演算をおこなっている部分を探すのは無理だった。
仮にできたとしても、環境依存なコードになる可能性が高い。

## まとめ

以上の検討を経て、64bit浮動小数で表現可能である32bit整数を`int`型とし、浮動小数もそれに合わせて32bitまでになるように実装をするのが現実的なのではないかという結論になった。

**Comment by Shinonome517** [comment]
BigIntのコンストラクタにSAFE.INTの範囲を超えた値を入れていたため、BigIntが任意精度でないと勘違いしていた。

BigIntを持ちいれば、64bit整数型を用意することが可能。

**Comment by Shinonome517** [comment]
まずは32bitで全部揃えて実装したのちに、余裕があれば64bitで実装する。

**Comment by narung56** [comment]
これDiscussionとしては消してよいと思う。
